extensions/mssql/src/telemetry/telemetry.ts

```diff
@@ -34,14 +34,80 @@ const telemetryReporter = new AdsTelemetryReporter<
     TelemetryActions | string
 >(packageInfo.name, packageInfo.version, packageInfo.aiKey);
 
+// Reusable error object for stack capture - avoids allocation overhead
+const stackCaptureError = { stack: "" };
+
+// Function names to skip in call stack (telemetry internals)
+const SKIP_FUNCTIONS = new Set([
+    "captureCallStack",
+    "sendActionEvent",
+    "sendErrorEvent",
+    "startActivity",
+    "update",
+    "end",
+    "endFailed",
+]);
+
+/**
+ * Captures a call stack with minimal overhead.
+ * Uses V8's captureStackTrace for fastest possible capture.
+ * Filters out telemetry internal functions.
+ */
+function captureCallStack(): string {
+    // V8 optimization: reuse object, skip this function in trace
+    Error.captureStackTrace(stackCaptureError, captureCallStack);
+    const raw = stackCaptureError.stack;
+
+    // Fast extraction using indexOf/slice - avoids regex overhead
+    let result = "";
+    let count = 0;
+    let pos = 0;
+
+    while (count < 8 && pos < raw.length) {
+        // Find "at "
+        const atPos = raw.indexOf("at ", pos);
+        if (atPos === -1) break;
+
+        // Find end of line
+        let lineEnd = raw.indexOf("\n", atPos);
+        if (lineEnd === -1) lineEnd = raw.length;
+
+        // Find opening paren (if exists) to know where function name ends
+        let nameEnd = raw.indexOf(" (", atPos + 3);
+        if (nameEnd === -1 || nameEnd > lineEnd) {
+            nameEnd = lineEnd;
+        }
+
+        const name = raw.slice(atPos + 3, nameEnd).trim();
+
+        // Skip empty, anonymous, Object. prefixed, and telemetry internal functions
```

Comment:
I don't have a better solution for it, but JS's proclivity for anonymous functions is probably going to result in a bunch of lost context here.